### ISSUE TITLE
Possible to elide more usages of Union types?

### ISSUE THREAD

**Issue by jgrund** [fable-compiler/ts2fable#109]
Hi, 

really nice work with the ts2fable update, it looks really nice.

Looking through the generated node bindings, I am wondering if we can elide more usage of Union types.

I get that they fit the TS definition usage nicely, but they become un-ergonomic to use when working with bindings and having to look at which case maps to which variant.

Would it be possible to collapse into method overloads instead? If we do that, then Union types (or a DU wrapper in general) become unneeded (can pass in type directly) + still get the benefits of type-checking signature.

**Comment by ctaggart** [comment]
Yes, it would be great if we could do this in combination with #110 to have uniquely named overloads. Now accepting design proposals.

**Comment by MangelMaxime** [comment]
We need to be careful with methods overloads. F# have difficulties determining the right overload as stated here https://github.com/fable-compiler/ts2fable/issues/110

**Comment by ctaggart** [comment]
One example is property setters. For this
``` fs
    type [<AllowNullLiteral>] TlsOptions =
        abstract host: string option with get, set
        abstract port: float option with get, set
        abstract pfx: U2<string, ResizeArray<Buffer>> option with get, set
        abstract key: U4<string, ResizeArray<string>, Buffer, ResizeArray<obj>> option with get, set
```

It would be nice to be able to set easily for each case:
``` fs
    type [<AllowNullLiteral>] TlsOptions =
        abstract host: string option with get, set
        abstract port: float option with get, set
        abstract pfx: U2<string, ResizeArray<Buffer>> option with get, set
        abstract pfxC1: string option with set
        abstract pfxC2: ResizeArray<Buffer> option with set
        abstract key: U4<string, ResizeArray<string>, Buffer, ResizeArray<obj>> option with get, set
        abstract keyC1: string option with set
        abstract keyC2: ResizeArray<string> option with set
        abstract keyC3: Buffer option with set
        abstract keyC4: ResizeArray<obj> option with set
```

**Comment by ctaggart** [comment]
For the unions in the functions, building off of a #110 proposal, we could add functions for each unique combination of cases.
``` fs
        abstract readFileSync: path: U2<PathLike, float> * ?options: obj option -> Buffer
        /// Synchronously reads the entire contents of a file.
        abstract readFileSync: path: U2<PathLike, float> * options: U2<obj, string> -> string
        /// Synchronously reads the entire contents of a file.
        abstract readFileSync: path: U2<PathLike, float> * ?options: U2<obj, string> option -> U2<string, Buffer>
```

would become
``` fs
        /// Synchronously reads the entire contents of a file.
        abstract readFileSync: path: U2<PathLike, float> * ?options: obj option -> Buffer
        abstract readFileSyncC1: path: PathLike * ?options: obj option -> Buffer
        abstract readFileSyncC2: path: float * ?options: obj option -> Buffer

        /// Synchronously reads the entire contents of a file.
        abstract readFileSyncP2: path: U2<PathLike, float> * options: U2<obj, string> -> string
        abstract readFileSyncP2C1C1: path: PathLike * options: obj -> string
        abstract readFileSyncP2C1C2: path: PathLike * options: string -> string
        abstract readFileSyncP2C2C1: path: float * options: obj -> string
        abstract readFileSyncP2C2C2: path: float * string -> string

        /// Synchronously reads the entire contents of a file.
        abstract readFileSyncP2N2: path: U2<PathLike, float> * ?options: U2<obj, string> option -> U2<string, Buffer>
        abstract readFileSyncP2N2C1C1: path: PathLike * ?options: obj option -> U2<string, Buffer>
        abstract readFileSyncP2N2C1C2: path: PathLike * ?options: string option -> U2<string, Buffer>
        abstract readFileSyncP2N2C2C1: path: float * ?options: obj option -> U2<string, Buffer>
        abstract readFileSyncP2N2C2C2: path: float * ?options: string option -> U2<string, Buffer>
```

Good/bad/ugly but useful?

**Comment by jgrund** [comment]
Hrrm, I haven't had many issues with method overloads. 

I think the key to that is avoiding any usage of types that can carry multiple meanings i.e. `obj` and `Function`.

Node is heavily based on method overloads, and I think the most common usage of obj is usually an options object (which can be defined as a record).

**Comment by jgrund** [comment]
@ctaggart using the `readFileSync` method above, if we know the option rules, we can encode all possible variants without issue using overloads. Here is an example that will allow every combination     (and know the return type) without the compiler complaining, and without any wrapper types:

[example](http://fable.io/repl/#?code=open%20System%0Aopen%20Fable.Core.JsInterop%0Aopen%20Fable.Core%0A%0Atype%20Buffer%20%3D%20Buffer%0Atype%20URL%20%3D%20URL%0A%0A%0Atype%20ReadFileSyncOpts%20%3D%20%7B%0A%20%20encoding%3A%20string%3B%0A%20%20flag%3A%20string%3B%0A%7D%0A%0A%0Atype%20ReadFileSyncOptsFlag%20%3D%20%7B%0A%20%20flag%3A%20string%3B%0A%7D%0A%0Atype%20IExports%20%3D%0A%20%20%2F%2F%2F%20Synchronously%20reads%20the%20entire%20contents%20of%20a%20file.%0A%20%20abstract%20readFileSync%3A%20path%3A%20string%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20float%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20Buffer%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20URL%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20string%20*%20encoding%3A%20string%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20float%20*%20encoding%3A%20string%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20Buffer%20*%20encoding%3A%20string%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20URL%20*%20encoding%3A%20string%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20string%20*%20options%3A%20ReadFileSyncOpts%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20float%20*%20options%3A%20ReadFileSyncOpts%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20Buffer%20*%20options%3A%20ReadFileSyncOpts%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20URL%20*%20options%3A%20ReadFileSyncOpts%20-%3E%20string%0A%20%20abstract%20readFileSync%3A%20path%3A%20string%20*%20options%3A%20ReadFileSyncOptsFlag%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20float%20*%20options%3A%20ReadFileSyncOptsFlag%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20Buffer%20*%20options%3A%20ReadFileSyncOptsFlag%20-%3E%20Buffer%0A%20%20abstract%20readFileSync%3A%20path%3A%20URL%20*%20options%3A%20ReadFileSyncOptsFlag%20-%3E%20Buffer%0A%0Alet%20Fs%3A%20IExports%20%3D%20jsNative%0A%0AFs.readFileSync%20%22foo%22%0AFs.readFileSync%203.0%0AFs.readFileSync%20URL%0AFs.readFileSync%20(%22foo%22%2C%20%22utf8%22)%0AFs.readFileSync(%22foo%22%2C%20%7B%20encoding%20%3D%20%22utf8%22%3B%20flag%20%3D%20%22r%22%20%7D)%0AFs.readFileSync(Buffer%2C%20%7B%20encoding%20%3D%20%22utf8%22%3B%20flag%20%3D%20%22r%22%20%7D)%0AFs.readFileSync(Buffer%2C%20%7B%20flag%20%3D%20%22r%22%20%7D)%0A)

<img width="961" alt="screen shot 2017-11-22 at 6 00 27 pm" src="https://user-images.githubusercontent.com/458717/33153333-4c036556-cfaf-11e7-9f5d-c96c6955f135.png">


We can do this by knowing the option contents, and because the node docs state: "If the encoding option is specified then this function returns a string. Otherwise it returns a buffer."

**Comment by ctaggart** [comment]
Ah, so the main problem is that ts2fable is generating `obj` for `TypeLiteral`s #93.

From node/index.d.ts, the 3 functions are:
``` ts
    export function readFileSync(path: PathLike | number, options?: { encoding?: null; flag?: string; } | null): Buffer;
    export function readFileSync(path: PathLike | number, options: { encoding: string; flag?: string; } | string): string;
    export function readFileSync(path: PathLike | number, options?: { encoding?: string | null; flag?: string; } | string | null): string | Buffer;
```

With the `TypeLiteral`s being:
``` ts
{ encoding?: null; flag?: string; }
{ encoding: string; flag?: string; }
{ encoding?: string | null; flag?: string; }
```

The problem here is that is duck typing. We could potentially find and interfaces that implement the types and add the interface to them, but let's cross that bridge when needed. I'm starting to thing that these types should be extracted. In this case though, it is 3 different types.

**Comment by jgrund** [comment]
Yes, extracting the options from the bindings is definitely the direction to move in I think.

That said, there are cases (this being one), where the TS bindings need updating to be less ambiguous.

The third case:

`export function readFileSync(path: PathLike | number, options?: { encoding?: string | null; flag?: string; } | string | null): string | Buffer;`

Doesn't specify anything concrete in either options or result, which is why I excluded it in my earlier example.

**Comment by alfonsogarciacaro** [comment]
I'm really enjoying this discussion. I'm very happy to see we're working together to make ts2fable a great tool! 🙌 

On erased union types, just a reminder that we have the `!^` operator (aka erased cast operator) that can make the profusion of erased union types in the bindings a bit more bearable. For example, for a setter:

```fsharp
    type [<AllowNullLiteral>] TlsOptions =
        abstract pfx: U2<string, IList<float>> option with get, set

    let opts: TlsOptions = jsOptions(fun o ->
        o.pfx <- !^"foo"
        o.pfx <- !^[|4., 5.|]
        // o.pfx <- !^6. // Doesn't compile, so we have type checking
    )
```

> About using `IList` instead of `ResizeArray` to represent JS arrays, see https://github.com/fable-compiler/ts2fable/issues/15#issuecomment-337854733

But I know the operator is still not ideal and some people don't like it. I added a suggestion so F# implements erased unions natively to make the operator redundant, please vote and/or comment there if you want to see it happen: https://github.com/fsharp/fslang-suggestions/issues/538

About using overloads in special places, or in general manually tweaking the bindings, I do agree this will probably still be helpful even if ts2fable reaches perfection (something that @ctaggart is bound to achieve! :wink:) if only because the TS bindings are not always perfect as @jgrund says. I've been wondering if there's a good way to _mark_ this manual changes so the tool won't touch that part (unless the corresponding TS part has changed, in which case a warning could be raised). Not sure if it'll be too complicated, in the last case we can just add a comment and use the git diffing to undo changes overwriting manual tweaks.

**Comment by jgrund** [comment]
> I've been wondering if there's a good way to mark this manual changes so the tool won't touch that part (unless the corresponding TS part has changed, in which case a warning could be raised).

I've been thinking about this as well. It seems like we should prefer updating the DefinitelyTyped bindings when we find they can be improved. That way, all future ts2fable generations of a lib would not need further updating.

Barring that, I guess the simplest approach would be to generate git .patch files and see if they apply cleanly. If they don't then someone can come in and resolve the conflicts.

Finally, we should probably have tests for the updated bindings like [DefinitelyTyped]( https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/node-tests.ts) to ensure we have a stable interface post applying any patches.

**Comment by alfonsogarciacaro** [comment]
Definitely, being able to translate also the tests for the declarations and run them would be a huge win 👍
